### PR TITLE
Update to Go 1.10 (and require that version)

### PR DESCRIPTION
**Pull request by jim**
## Description

Our project depends on Go 1.9.3 currently. I'd like to prevent confusion and issues that stem from version incompatibilities.

_Note: This PR has been updated to require Go 1.10.x._

## Reviewer Notes

* While Go has a policy of backwards compatibility for Go 1.x and therefore our code should continue to work fine on newer versions, _this guarantee does not extend to the go tooling or any tools we use that are built on top of them_. One example of this is that Go 1.10 made some changes to `go vet` that broke one of our pre-commit hooks. We are also dependent upon several third party libraries, the version compatibilities of which we do not track.
* [prereqs](https://github.com/trussworks/prereqs) considers that specific tool versions should be guaranteed as part of a project:
    > prereqs only tests for a tools existence in the PATH (or as a shell function); it doesn't try to check versions. If you want that you should consider installing your tool as part of your project.
* We could use one of the Go version managers, but I don't love the idea of our project depending on one of those to run. My opinion is that the project should verify that it is running with the correct Go version (and potentially provide hints on how to get that setup), but that managing the versions of Go installed on a machine is not a project-level concern.
* I considered having the script just print a warning, but I think that it is likely to be ignored.
* I am far from a bash or shell master and would appreciate any feedback on this PR!

## Verification Steps

* [ ] The bash code is reasonable and understandable.
* [ ] The code prevents running the project on the wrong version of Go.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/155346020) for this change

### CODE REVIEW

**Comment by jim** [comment]
@brainsik Including you as this may affect CI or deployment tooling.

File: bin/check_go_version [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#! /bin/bash
+
+set -eu -o pipefail
+
+VERSION="go1.9.3"
```
Comment (brainsik) [inline bin/check_go_version]: If we trust that patch versions are generally safe and backwards compatible, we could relax this to only check for major.minor. That might ease up on some worries about getting blocked by a `brew upgrade`.

**Comment by macrael** [comment]
Sounds good to me. Another thing I like about this strategy is that it forces us to keep everyone updating as new minor releases are released, just like we are updating our dependencies regularly.

**Comment by jim** [comment]
Thanks for the input, everyone.

* Based on the conversation, I'll update the requirement to `go1.10` , which should allow any patch versions to match. 
* I agree that we should consider how difficult it is to run the project under other versions, but to me, that's more of an issue of using `brew` to manage `go` than anything with the project. I'll defer resolving this for another task.

**Comment by jim** [comment]
@brainsik LMK if there is anything I can do to get the primary image updated. Otherwise, we can put this on hold until that is done.

**Comment by brainsik** [comment]
@jim To update to the latest primary image:
1. Gab the latest tag from here: https://microbadger.com/images/trussworks/circleci-docker-primary
1. Update the CircleCI config to use that tag

Let me know if you run into trouble.

**Pull request merged**